yw_website/yw_website/apps/website/serializers.py

```diff
@@ -157,25 +283,180 @@ def __create_tags(self, w, validated_data):
+
+    def _create_files(self, r, validated_data):
+        files = FileSerializer(validated_data.get("files"), many=True)
         for file in files.data:
             f, _ = File.objects.update_or_create(
-                checksum=file.get('checksum'),
-                defaults={                
-                    'size':file.get('size'),
-                    'name':file.get('name'),
-                    'uri':file.get('uri'),
-                    'last_modified':file.get('lastModified')
-                }
+                checksum=file.get("checksum"),
+                defaults={
+                    "size": file.get("size"),
+                    "name": file.get("name"),
+                    "uri": file.get("uri"),
+                    "last_modified": file.get("lastModified"),
+                },
             )
             f.save()
             rf = RunFile(run=r, file=f)
             rf.save()
 
+    def _create_program_blocks(self, r, validated_data):
+        program_blocks = ProgramBlockSerializer(
+            validated_data.get("programBlock"), many=True, context={"run": r}
+        )
+        for program_block in program_blocks.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=program_block.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            pb = ProgramBlock(
+                programblock_id=program_block.get("programBlockId"),
+                name=program_block.get("name"),
+                qualified_name=program_block.get("qualifiedName"),
+                in_program_block=in_block,
+                run=r,
+            )
+            pb.save()
+
+    def _create_data(self, r, validated_data):
+        datas = DataSerializer(
+            validated_data.get("data"), many=True, context={"run": r}
+        )
+        for data in datas.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=data.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            d = Data(
+                data_id=data.get("dataId"),
+                name=data.get("name"),
+                qualified_name=data.get("qualifiedName"),
+                in_program_block=in_block,
+                run=r,
+            )
+            d.save()
+
+    def _create_ports(self, r, validated_data):
+        ports = PortSerializer(
+            validated_data.get("port"), many=True, context={"run": r}
+        )
+        for port in ports.data:
+            try:
+                in_block = ProgramBlock.objects.get(
+                    programblock_id=port.get("inProgramBlock"), run=r
+                )
+            except ProgramBlock.DoesNotExist:
+                in_block = None
+
+            try:
+                data = Data.objects.get(data_id=port.get("data"), run=r)
```

Comment:
there should always be a valid data_id associated with a port, if thats not the case then we should throw a 500 because we received bad input.